### PR TITLE
Feature/allow multiple users cloud sim

### PR DESCRIPTION
**Pull request by adev4a**
# PR Details
This PR updates the cloud-sim deployment to allow multiple users to start cloud sessions for XiL.

## Description
All CARMA cloud sim sessions are hosted under a single VPC. Each successive user is added to this VPC under a new subnet.

The cloud-sim bash script is updated to add queries through aws cli to check if the VPC already exists. If it doesn't a VPC, Internet gateway and route table is created. In case it does exist, the subnet value is updated by 1 and a new user is created with it.

## Types of changes

- [x] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by kjrush — CHANGES_REQUESTED** [review]

File: engineering_tools/carma_script_extensions/configure-cloud-sim.bash [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    igw_id=$(aws ec2 create-internet-gateway \
+    --tag-specifications 'ResourceType=internet-gateway,Tags=[{Key=Name,Value=XiL-Cloud IGW}]' --query 'InternetGateway.InternetGatewayId' \
+    --output text) 
+
+    # Check the exit code to determine if the Internet Gateway creation was successful
+    if [ $? -eq 0 ]; then
+        igw_success=1
+        echo "Created Internet Gateway: $igw_id"
+    else
+        echo "Error creating Internet Gateway: $igw_id"
```
Comment (kjrush) [inline engineering_tools/carma_script_extensions/configure-cloud-sim.bash:78]: I think in each failure case here, we should clean up all the resources we'd allocated before the failure, then exit the script so we don't allocate any more. Ideally the only possible outcomes of this script are "successful fully initialized AWS resources" or "nothing happened".

File: engineering_tools/carma_script_extensions/configure-cloud-sim.bash [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+    if [ $vpc_success -eq 1 ] && [ $igw_success -eq 1 ] && [ $route_table_success -eq 1 ]; then
+        echo "XiL-Cloud VPC, Internet Gateway and Route Table were deleted successfully!"
+    else
+        echo "No resources required to be deleted"
+    fi
+
+}
+
+configure-cloud-sim__help() {
```
Comment (kjrush) [inline engineering_tools/carma_script_extensions/configure-cloud-sim.bash:175]: Is this ever used? only "display_help" is referenced below.

File: engineering_tools/carma_script_extensions/cloud-sim [hunk 3]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-#  Copyright (C) 2022 LEIDOS.
+#  Copyright (C) 2022-2023 LEIDOS.
```
Comment (kjrush) [inline engineering_tools/carma_script_extensions/cloud-sim:3]: The help output for this script should reference the other configure-cloud-sim.bash to tell users that one needs to be run first,  I think.

**Review by kjrush — CHANGES_REQUESTED** [review]

File: engineering_tools/carma_script_extensions/configure-cloud-sim.bash (hunk 1 above)
Comment (adev4a) [reply engineering_tools/carma_script_extensions/configure-cloud-sim.bash:78]: Yup thats a good point. In a case where we failed to set up one of the resources, configuration has faile

File: engineering_tools/carma_script_extensions/cloud-sim [hunk 4]
```diff
@@ -158,104 +158,164 @@ cloud-sim__start() {
     
     # Call cloud formation
     local STACK_NAME="CloudSimStack"$(echo ${USER_ARN} |  grep -oh "/.*@" | sed "s|@|-|g" | sed "s|\.|-|g" | sed "s|[^A-Za-z0-9]|-|g")
-
```
Comment (adev4a) [inline engineering_tools/carma_script_extensions/cloud-sim:161]: Just to simplify this a little, none of the logic in red here has been removed. These calls have moved to a different line.
In the green, Lines 214-314 have just been indented - no change in code

File: engineering_tools/carma_script_extensions/cloud-sim [hunk 5]
```diff
@@ -217,112 +209,113 @@ cloud-sim__start() {
...
+        echo "STACK_NAME=${STACK_NAME}" > ${CARMA_CLOUD_SIM_STACK_PATH}
+        
+        # Print for user
+        echo "##############################"
+        echo ""
+        echo "AWS Stack ${STACK_NAME} Launch Suceeded"
+        echo ""
+        echo "EC2 Instance Address: ${PUBLIC_DNS}"
+        echo ""
+        echo "Cloud9 IDE Address: ${CLOUD9_ADDRESS}"
```
Comment (kjrush) [inline engineering_tools/carma_script_extensions/cloud-sim:311]: Out of curiosity, did you confirm that the Cloud9 config still works? Not sure it's super high priority, but it is something that's nice for development.

File: engineering_tools/carma_script_extensions/cloud-sim [hunk 6]
```diff
@@ -358,6 +351,211 @@ cloud-sim__status() {
...
+    
+    # Check if resources already exist - reset if all don't exist
+    local vpc_success=0
+    local igw_success=0
+    local route_table_success=0
+
+    # Get resource status - query_resource_status defines VPC, Internet Gateway and Route Table ID
+
+    if query_resource_status; then
+        echo "Xil Cloud resources already exist. No change required. Run start to bring up stack.
```
Comment (kjrush) [inline engineering_tools/carma_script_extensions/cloud-sim:376]: It's not a huge deal, but I would recommend where possible we change "XIL" to "CDASim" here, XIL is the project name, but the product is CDASim.

Comment (kjrush) [reply engineering_tools/carma_script_extensions/cloud-sim:376]: Also would be good to change "start" to "carma cloud-sim start" so it's unambiguous, for users that haven't done this before.

File: engineering_tools/carma_script_extensions/cloud-sim [hunk 7]
```diff
@@ -387,7 +585,10 @@ Please enter one of the following commands when using the cloud-sim extension:
                 -i 
                     - Specify the exact carma-config to use for launching the vehicle and simulation
                       For example " -i image usdotfhwastol/carma-config:carma-system-1.0.1 "
-                      This option is an alternative to -v and should only be used if that option is not specified. 
+                      This option is an alternative to -v and should only be used if that option is not specified.
+            init
```
Comment (kjrush) [inline engineering_tools/carma_script_extensions/cloud-sim:589]: Just for consistency sake, maybe this is better called "init-resources" if the inverse is called "cleanup-resources"?

File: engineering_tools/carma_script_extensions/cloud-sim (hunk 5 above)
Comment (adev4a) [reply engineering_tools/carma_script_extensions/cloud-sim:311]: I just tested this and it gave me an error trying to access Cloud9
"1 validation error detected: Value 'ide' at 'environmentId' failed to satisfy constraint: Member must satisfy regular expression pattern: ^[a-zA-Z0-9]{8,32}$"

Comment (adev4a) [reply engineering_tools/carma_script_extensions/cloud-sim:311]: Issue logged here https://github.com/usdot-fhwa-stol/carma-platform/issues/2169

**Review by kjrush — APPROVED** [review]

**Pull request merged**